File: bound/kt/src/main/kotlin/web5/sdk/dids/BearerDid.kt
package web5.sdk.dids

import web5.sdk.crypto.keys.KeyManager
import web5.sdk.crypto.keys.ToOuterKeyManager
import web5.sdk.crypto.signers.Signer
import web5.sdk.crypto.signers.ToOuterSigner
import tbdex.sdk.rust.DidData as RustCoreDid
import tbdex.sdk.rust.DocumentData as RustCoreDocument
import tbdex.sdk.rust.BearerDid as RustCoreBearerDid

typealias Did = RustCoreDid
typealias Document = RustCoreDocument

class BearerDid {
    val did: Did
    val document: Document
    val keyManager: KeyManager

    internal val rustCoreBearerDid: RustCoreBearerDid

    constructor(uri: String, keyManager: KeyManager) {
        this.rustCoreBearerDid = RustCoreBearerDid(uri, keyManager.getRustCoreKeyManager())

        this.did = this.rustCoreBearerDid.getData().did
        this.document = this.rustCoreBearerDid.getData().document
        this.keyManager = keyManager
    }

    constructor(portableDid: PortableDid) {
        this.rustCoreBearerDid = RustCoreBearerDid.fromPortableDid(portableDid.rustCorePortableDid)

        this.did = this.rustCoreBearerDid.getData().did
        this.document = this.rustCoreBearerDid.getData().document
        this.keyManager = ToOuterKeyManager(this.rustCoreBearerDid.getData().keyManager)
    }

    fun getSigner(): Signer {
        // TODO currently hardcoding to first VM
        val keyId = this.document.verificationMethod.first().id
        val innerSigner = this.rustCoreBearerDid.getSigner(keyId)
        return ToOuterSigner(innerSigner)
    }
}

File: bound/typescript/src/bearer-did.ts
import { withError } from "./errors";
import { KeyManager } from "./key-managers";
import { PortableDid } from "./portable-did";
import { Signer } from "./signers";
import wasm from "./wasm";
import { Did, Document } from "./wasm/mappings";

export class BearerDid {
  readonly did: Did;
  readonly document: Document;
  readonly keyManager: KeyManager;

  constructor(did: Did, document: Document, keyManager: KeyManager) {
    this.did = did;
    this.document = document;
    this.keyManager = keyManager;
  }

  static fromWASM = withError(
    (wasmBearerDid: wasm.WasmBearerDid): BearerDid => {
      return new BearerDid(
        Did.fromWASM(wasmBearerDid.did),
        Document.fromWASM(wasmBearerDid.document),
        KeyManager.fromWASM(wasmBearerDid.key_manager)
      );
    }
  );

  toWASM = withError((): wasm.WasmBearerDid => {
    return new wasm.WasmBearerDid(
      Did.toWASM(this.did),
      Document.toWASM(this.document),
      KeyManager.toWASM(this.keyManager)
    );
  });

  static fromPortableDID = withError((portableDID: PortableDid): BearerDid => {
    return BearerDid.fromWASM(
      wasm.WasmBearerDid.from_portable_did(portableDID.toWASM())
    );
  });

  getSigner = withError((verificationMethodId: string): Signer => {
    return Signer.fromWASM(this.toWASM().get_signer(verificationMethodId));
  });
}
